### PR TITLE
Implement `chainerx::SoftmaxCrossEntropy` and `chainerx.softmax_cross_entropy`

### PR DESCRIPTION
**Pull request by takagi**
This PR implements `chainerx::SoftmaxCrossEntropy` and `chainerx.softmax_cross_entropy`, as well as replacing its temporary implementations in Chainer inter-op and ChainerX MNIST example.

Limitations:
- The number of dimensions of `x1` is two only
- No reduction mode
- `ignore_label` can not be specified
- No `cache_score`, `class_weight` and `enable_double_backprop` arguments that `F.softmax_cross_entropy` has

I fixed `F. SoftmaxCrossEntropy`'s `forward_chainerx()` method to use `chainerx.softmax_cross_entropy` as well.

I will fix `chainer/chainerx_cc/examples/mnist/` and `chainer/chainerx_cc/examples/mnist_py/` to use the new ChainerX implementation in another PR.

### CODE REVIEW

File: chainerx_cc/chainerx/routines/loss.h [hunk 1]
```diff
@@ -15,6 +15,17 @@ Array HuberLoss(const Array& x1, const Array& x2, Scalar delta);
...
+enum class SoftmaxCrossEntropyReduceMode {
+    kMean = 1,
+    kNo,
+};
+
+Array SoftmaxCrossEntropy(
+        const Array& x1,
+        const Array& x2,
+        bool normalize = true,
+        SoftmaxCrossEntropyReduceMode reduce_mode = SoftmaxCrossEntropyReduceMode::kMean);
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/loss.h]: Following [this interface design principle](https://github.com/chainer/chainer/wiki/Development-policies#api-1), we can probably skip the `normalize` and `reduce_mode` altogether?

Comment (takagi) [reply chainerx_cc/chainerx/routines/loss.h]: Sure, I removed them!

**Comment by takagi** [comment]
@hyv I've made this PR ready with updating PR description. Would you start reviewing?

File: chainer/functions/loss/softmax_cross_entropy.py [hunk 2]
```diff
@@ -100,7 +100,7 @@ def _is_chainerx_supported(self, input_arrays):
             return False
         if self.ignore_label != -1:
             return False
-        if self.reduce != 'mean':
+        if self.reduce == 'mean' and self.normalize:
             return False
```
Comment (hvy) [inline chainer/functions/loss/softmax_cross_entropy.py]: These are the default configurations. How about instead of falling back, run the previous implementation in this case? The denominator https://github.com/chainer/chainer/pull/8250/files#diff-7f9fda8b243258115ef7bdac145fec8dL122 should be `mask.sum()` instead of `x.shape[0]` though.

Comment (hvy) [reply chainer/functions/loss/softmax_cross_entropy.py]: There was a bug with `normalize` which I fixed in https://github.com/chainer/chainer/pull/8301. Maybe we should wait for that one to be merged before.

File: tests/chainerx_tests/unit_tests/routines_tests/test_loss.py [hunk 3]
```diff
@@ -172,6 +172,42 @@ def forward_chainer(self, inputs):
         return out,
 
 
+@op_utils.op_test(['native:0'])
```
[the remark this answers is not in the document]
Comment (takagi) [reply tests/chainerx_tests/unit_tests/routines_tests/test_loss.py]: Added!

File: chainer/functions/loss/softmax_cross_entropy.py (hunk 2 above)
Comment (takagi) [reply chainer/functions/loss/softmax_cross_entropy.py]: Sounds good. I'll review #8301.

**Comment by hvy** [comment]
@takagi thanks for reviewing #8301. Could you resolve the conflicts in this PR that now occurred?

**Comment by takagi** [comment]
Resolved the conflict!

File: chainer/functions/loss/softmax_cross_entropy.py [hunk 4]
```diff
@@ -109,22 +109,19 @@ def _is_chainerx_supported(self, input_arrays):
...
-        score = chainerx.log_softmax(x, axis=1)
-        mask = (t[:, chainerx.newaxis] == chainerx.arange(
-            n_classes, dtype=t.dtype, device=x.device)).astype(score.dtype)
+        if self.reduce == 'mean' and self.normalize:
+            x, t = inputs
+            n_classes = x.shape[1]
+            score = chainerx.log_softmax(x, axis=1)
+            mask = (t[:, chainerx.newaxis] == chainerx.arange(
+                n_classes, dtype=t.dtype, device=x.device)).astype(score.dtype)
+            y = -(score * mask).sum() * (1 / mask.sum())
```
Comment (hvy) [inline chainer/functions/loss/softmax_cross_entropy.py]: One less routine call.
```suggestion
            y = (score * mask).sum() * (-1 / mask.sum())
```

Comment (takagi) [reply chainer/functions/loss/softmax_cross_entropy.py]: Nice!

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
LGTM! 

I'm thinking if we should introduce arguments after all similar to SCE in Chainer, simply to make it more useful. But this also applied to `SigmoidCrossEntropy` so we should probably address them both in the discussion.
